### PR TITLE
Improved configuration compiler

### PR DESCRIPTION
**Pull request by jmanero-r7**
The Config module now detects dirty state and recompiles itself until fully converged. This allows DSL files to reference `Config` freely with arbitrary orders of circular reference.

There is currently a default limit of 4 recompile iterations. `Config.compile` will throw an error if the `max_iterations` value is exceeded.

### CODE REVIEW

File: lib/builderator/config/attributes.rb [hunk 1]
```diff
@@ -102,8 +102,7 @@ def collection(collection_name, &definition)
       include Enumerable
 
       ## Delegate enumerables to underlying storage structure
-      def_delegators :@attributes,
-                     :[], :[]=, :fetch, :store,
+      def_delegators :@attributes, :[], :fetch, :[]=,
```
Comment (fmitchell-r7) [inline lib/builderator/config/attributes.rb]: Was `:store` intentionally removed from this list?

Comment (jmanero-r7) [inline lib/builderator/config/attributes.rb]: Yeah, it's not currently used at all, and we don't actually want to expose setters on the storage structure. There should already be named setters for any valid attribute. There are a few places where `[]=` is still used, which is why it's still lingering...

File: lib/builderator/control/version.rb [hunk 2]
```diff
@@ -36,9 +36,7 @@ def current
 
         def set_config_version
           Config.defaults.version = current.to_s
-
-          Config.recompile ## Update compiled version attribute
-          Config.recompile ## Pass new attribute back out to other layers
+          Config.recompile
```
Comment (rvanburen-r7) [inline lib/builderator/control/version.rb:39]: It looks like you changed all the recompiles to compile except this one. Was it meant to be left unchanged?

File: lib/builderator/config/rash.rb [hunk 3]
```diff
@@ -53,7 +53,7 @@ def merge!(other)
           k = k.gsub(/\-/, '_') if k.is_a?(String)
 
           ## Merge Arrays
-          next self[k] += v if fetch(k, nil).is_a?(Array) && v.is_a?(Array)
+          next self[k] |= v if fetch(k, nil).is_a?(Array) && v.is_a?(Array)
```
Comment (fmitchell-r7) [inline lib/builderator/config/rash.rb:56]: Is it worth adding some tests around a `Rash` to validate this sort of behavior?

File: lib/builderator/control/version.rb (hunk 2 above)
Comment (jmanero-r7) [inline lib/builderator/control/version.rb:39]: `recompile`'s an alias to `compile` right now. It's just semantics...

**Comment by rvanburen-r7** [comment]
LGTM

File: lib/builderator/config/attributes.rb (hunk 1 above)
Comment (fmitchell-r7) [inline lib/builderator/config/attributes.rb]: :+1:

File: lib/builderator/config/rash.rb (hunk 3 above)
Comment (jmanero-r7) [inline lib/builderator/config/rash.rb:56]: Possibly? `|=` of `Array` is set-union. The current expected behavior is that array-attributes will be sets of unique values. (e.g. run-lists, security-group lists)

Comment (jmanero-r7) [inline lib/builderator/config/rash.rb:56]: The issue was that during re-compiles, elements were just being re-appended to list-attributes

**Pull request merged**